### ISSUE TITLE
Out of memory on writing to tantivy index

### ISSUE THREAD

**Issue by VsevolodZakharov** [quickwit-inc/tantivy#1563]
I have "ValueError: Failed to open file for read: 'IoError { io_error: Os { code: 12, kind: OutOfMemory, message: \"Cannot allocate memory\" }, filepath: \"/index/TANTIVY/Content.en/LearningObjectives/9bab6227e72543aabb95be37a875fa77.fast\" when trying to write new data to the index. Increasing container's RAM does not help. Does writer memory limit influence that, or this file reading is independent?

**Comment by PSeitz** [comment]
Thanks for the bug report. The memory budget you give to the tantivy writer is not the upper limit, we go higher than that when flushing the index to disk (Should be around 2x if I remember correctly) . Are you using multiple threads when indexing? How much memory did you allocate? Maybe try lowering it.

Can you share your index config?

Getting the OOM while opening a file seems a little odd, since the allocation for opening a file should be dwarfed by tantivy's allocations.

**Comment by VsevolodZakharov** [comment]
Thanks for reply! That is how it is done (FYI:That is python binding):
```
def makeSchema(self):
        return SchemaBuilder()\
            .add_text_field("path", stored = True, index_option = "basic",  tokenizer_name = "raw")\
            .add_text_field("title", stored = "title" in self.storedFields, tokenizer_name = "default")\
            .add_text_field("content", stored = "content" in self.storedFields, tokenizer_name = "default")\
            .add_text_field("keywords", stored = "keywords" in self.storedFields, index_option = "basic", tokenizer_name = "whitespace")\
            .build()
    WRITER_HEAP_MB = 256
    WRITER_THREADS = 4
    w = self.getIndex().writer(heap_size=TantivyIndex.WRITER_HEAP_MB*1024*1024,
num_threads=TantivyIndex.WRITER_THREADS)
```

**Comment by PSeitz** [comment]
Can you test it with much less memory per indexer, e.g.

```
WRITER_HEAP_MB = 128
WRITER_THREADS = 1
```

**Comment by VsevolodZakharov** [comment]
Let me try, although the container had 16GM of RAM. Even 256Mb*2 is far from that 16Gb

**Comment by VsevolodZakharov** [comment]
Nothing helps. @PSeitz, is this correct place for that *.fast file openning error source: https://github.com/quickwit-oss/tantivy/blob/main/src/directory/directory.rs#L124  ?

**Comment by PSeitz** [comment]
Yes, that's the place. Opening a file doesn't cost barely any memory normally, do you use some sort of ramdisk or custom `Directory` implementation?

**Comment by fulmicoton** [comment]
Which directory are you using?

If it is MMapDirectory, keep in mind this is not necessarily a problem of lack of memory. You might be hitting a virtual memory limit. 
tantivy relies on mmap for its IO, and requires a lot of virtual memory. 
Make sure you do not have any limit on that.

Use your OS tools to inspect and report the resident memory usage and the virtual memory usage, to check which one is hitting your limit...

**Comment by PSeitz** [comment]
For the virtual memory size, can you post the output of
```
ulimit -a | grep virtual

cat /proc/meminfo | grep Vm
```

**Comment by VsevolodZakharov** [comment]
@fulmicoton I'm using this python binding : https://github.com/quickwit-oss/tantivy-py
And it looks like ut uses MmapDiractory: https://github.com/quickwit-oss/tantivy-py/blob/master/src/index.rs#L179

**Comment by VsevolodZakharov** [comment]
@PSeitz 
root@5aff4ff9ca15:/# ulimit -a | grep virtual
virtual memory              (kbytes, -v) unlimited
root@5aff4ff9ca15:/# cat /proc/meminfo | grep Vm
VmallocTotal:   34359738367 kB
VmallocUsed:       90584 kB
VmallocChunk:          0 kB

**Comment by VsevolodZakharov** [comment]
@PSeitz , is ManagedDirectory an alternative for MmapDirectory? Or is there any way to avoid mmap-ing?

**Comment by fulmicoton** [comment]
No.
